runners/google-cloud-dataflow-java/examples/build.gradle

```diff
@@ -97,6 +97,42 @@ task verifyPortabilityApi() {
   dependsOn verifyFnApiWorker
 }
 
+def passedDockerImageName = project.findProperty('passedDockerImageName')
+def java8Home = project.findProperty('java8Home')
+def java11Home = project.findProperty('java11Home')
+
+task runFnApiTest(type: Test) {
+    def dataflowWorkerJar = project.findProperty('dataflowWorkerJar') ?: project(":runners:google-cloud-dataflow-java:worker").shadowJar.archivePath
+    with commonConfig(dataflowWorkerJar, passedDockerImageName, ["--experiments=${fnapiExperiments}"])
+    useJUnit {
+        excludeCategories 'org.apache.beam.sdk.testing.StreamingIT'
+    }
+}
+
+task verifyFnApiWorkerJava11(type: Test) {
+    dependsOn ":runners:google-cloud-dataflow-java:worker:shadowJar"
+    dependsOn ":runners:google-cloud-dataflow-java:buildAndPushDockerContainer"
+    doLast {
+        exec {
+            workingDir "${rootProject.projectDir}"
+            commandLine "./gradlew", "${project.path}:runFnApiTest", "-Pjava8Home=${java8Home}", "-PcompileWithJava8", "-PpassedDockerImageName=${dockerImageName}", "-x shadowJar", "-x shadowTestJar", "-x jar", "-x testJar", "-x classes", "-x testClasses", "-PdisableSpotlessCheck=true", "-Dorg.gradle.java.home=${java11Home}"
```

Comment:
I think baking the exclusions and other stuff into Gradle is worse than doing it explicitly from Jenkins job definition. What is the advantage of this approach?

Reply:
Intention is to compile test classes with java 8 and then run them with java 11. My first implementation was based on solution written and already used in project. So far if we want to run gradle task with other jdk version then default we use `-Dorg.gradle.java.home=[PATH_TO_JAVA_HOME]`. Since I failed to find another way I did the same. Then in turned out, in this particular case, I need to pass docker image name to second gradle task (one that should run with java 11). I managed to do so by invoking another gradle task inside. We use similar mechanism in other job scripts.
But I think we are right, I can remove all `-x` flags since none of them is related with compilation task.

Reply:
Ok, I understand it now. It's not an ideal solution but it will have to suffice

Reply:
Note that the copying the existing ways doesn't remove how they are suboptimal since Gradle may some day choose to overwrite the output done by the past "partial" executions because it detected and was told to run using a different configuration.

Also, executing Gradle in the same directory another Gradle is currently running may corrupt the build output of each since the output could be written by two or more tasks.

Reply:
@lukecwik thanks for input!
Before you proceed with my comment just keep in mind I am not gradle ninja :)
I am totally aware it is not best/good-looking solution. TBH, it's the only one I figured out.
I've checked and "main" task is waiting till sub task is finished, then it proceeds.`runFnApiTest` runs as last one (well `cleanUpDockerImages` runs as last) in this jenkins job. But my point is, it's the last (and only) task that produces output (results from tests).
I am doing my best but please, feel free to give me some guidance if you have an idea how this could be accomplished.

Reply:
@Ardagan can you please help out?

If you want to use a single gradle command, you typically need to build different variants as separate artifacts. e.g. produce Java 11 classes/jar and Java 8 classes/jar and use the correct JVM when doing so in the compileJava and compileTestJava tasks (and ensure that the [JavaCompile](https://docs.gradle.org/current/dsl/org.gradle.api.tasks.compile.JavaCompile.html) task is configured to use the correct JVM). For running tests, you have to configure the [Test](https://docs.gradle.org/current/dsl/org.gradle.api.tasks.testing.Test.html) to use the correct JVM.

Anything that relies on independent gradle executions needs to perform all the artifact copying/moving between executions to ensure that output isn't overwritten/corrupted between Gradle runs.

Reply:
Since you provided those links... is it as simple as creating duplicate test tasks that specify `targetCompatibility` for test compile and `executable` for the tests?

Reply:
@kennknowles I also need to pass proper docker image name. That is the reason why I did it this way

Reply:
> @Ardagan can you please help out?
> 
> If you want to use a single gradle command, you typically need to build different variants as separate artifacts. e.g. produce Java 11 classes/jar and Java 8 classes/jar and use the correct JVM when doing so in the compileJava and compileTestJava tasks (and ensure that the [JavaCompile](https://docs.gradle.org/current/dsl/org.gradle.api.tasks.compile.JavaCompile.html) task is configured to use the correct JVM). For running tests, you have to configure the [Test](https://docs.gradle.org/current/dsl/org.gradle.api.tasks.testing.Test.html) to use the correct JVM.
> 
> Anything that relies on independent gradle executions needs to perform all the artifact copying/moving between executions to ensure that output isn't overwritten/corrupted between Gradle runs.

At line L123 I set javaHome for `AbstractCompile` tasks. 
I forgot to add link jenkins job that tested this solution [JOB](https://builds.apache.org/job/beam_PreCommit_JavaPortabilityApiJava11_Commit/20/)

Reply:
@kennknowles @lukecwik As I see it, current implementation is no-go for you and I should close this PR, am I correct?

Reply:
@lukecwik 

> Anything that relies on independent gradle executions needs to perform all the artifact copying/moving between executions to ensure that output isn't overwritten/corrupted between Gradle runs.

This is why the second Gradle command is started with explicit exclusions of all compilation tasks. The worst-case scenario is that somehow the built sources aren't good enough and the task fails. 

We researched multiple ways of doing what is done in this PR - in a nutshell, Gradle doesn't seem to be well suited for JVM manipulations in subsequent stages of a task. Best we could do was a hack like this one or like in [job_PostCommit_Java_ValidatesRunner_Dataflow_Java11.groovy](https://github.com/apache/beam/blob/master/.test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Dataflow_Java11.groovy).

If you know of a way of such manipulation that would be more elegant, please let us know - we're painfully aware that this is suboptimal and error-prone. Still, to my knowledge it is still the only way to make it work.
Also tagging @kennknowles - if you know something, please share :)